eth/consensus/clique/clique.py

```diff
@@ -0,0 +1,173 @@
+)
+
+
+def configure_header(vm: VirtualMachineAPI, **header_params: HeaderParams) -> BlockHeaderAPI:
+    with vm.get_header().build_changeset(**header_params) as changeset:
+        # We do this because the default `configure_header` tries to compute the difficulty
+        # which we can not do at this point. We take the `difficulty` as provided and rely
+        # on the `validate_seal` call that will happen as the last step when blocks are
+        # imported.
+        header = changeset.commit()
+    return header
+
+
+def _construct_turn_error_message(expected_difficulty: int,
+                                  header: BlockHeaderAPI,
+                                  signer: Address,
+                                  signers: Sequence[Address]) -> str:
+
+    return (
+        f"Expected difficulty of {header} to be {expected_difficulty} "
+        f"but was {header.difficulty}.\n"
+        f"Header signer: {encode_hex(signer)}. Valid signers: {signers}"
+    )
+
+
+class CliqueConsensus:
+    """
+    This class is the entry point to operate a chain under the rules of Clique consensus which
+    is defined in EIP-225: https://eips.ethereum.org/EIPS/eip-225
+    """
+
+    logger = logging.getLogger('eth.consensus.clique.CliqueConsensus')
+
+    def __init__(self, base_db: AtomicDatabaseAPI, epoch_length: int = EPOCH_LENGTH) -> None:
+        if base_db is None:
+            raise ValueError("Can not instantiate without `base_db`")
+        self._epoch_length = epoch_length
+        self._chain_db = ChainDB(base_db)
+        self._header_cache = HeaderCache(self._chain_db)
+        self._snapshot_manager = SnapshotManager(
+            self._chain_db,
+            self._header_cache,
+            self._epoch_length,
+        )
+
+    @to_tuple
+    def amend_vm_configuration(self, config: VMConfiguration) -> Iterable[VMFork]:
+        """
+        Amend the given ``VMConfiguration`` to operate under the rules of Clique consensus.
+        """
+        for pair in config:
+            block_number, vm = pair
+            vm_class = vm.configure(
+                extra_data_max_bytes=65535,
+                validate_seal=staticmethod(self.validate_seal),
+                create_execution_context=staticmethod(self.create_execution_context),
+                configure_header=configure_header,
+                _assign_block_rewards=lambda _, __: None,
+            )
+
+            yield block_number, vm_class
+
+    @staticmethod
+    def create_execution_context(header: BlockHeaderAPI,
+                                 prev_hashes: Iterable[Hash32],
+                                 chain_context: ChainContext,
+                                 import_block: BlockHeaderAPI=None) -> ExecutionContext:
+
+        # In Clique consensus, the tx fee goes to the signer
+        coinbase = header.coinbase
+        if import_block is not None:
+            try:
+                coinbase = get_block_signer(import_block.header)
```

Comment:
This is a drive-by thought without reading the full PR: maybe there could be a special header class where `header.coinbase = get_block_signer(header)`. But I'm also missing something about why the block signer can't be pulled from the original header.

Reply:
I'm :+1: on moving this direction but lets save it for future work.  This would involve some level of plumbing across the `ChainAPI` and `VirtualMachineAPI` to allow specifying custom header classes.

Reply:
>But I'm also missing something about why the block signer can't be pulled from the original header.

The `header` in `create_execution_context` wasn't the *original* header but the *in progress* header that we keep on the vm. I changed that now to pass in the one that is passed into `import_block` so that we got rid of that extra parameter.